Initial documentation for Environments (#17096)
pantsbuild/pants: docs/markdown/Using Pants/environments.md

Environment targets are given short, descriptive names using the [`[environments.names]` option](doc:reference-environments#names) (usually defined in `pants.toml`), which consuming targets use to refer to them in `BUILD` files. That might look like a `pants.toml` section and `BUILD` file (at the root of the repository in this case) containing:

```toml
[environments.names]
linux = "//:local_linux"
linux_docker = "//:local_busybox"
```

```python
local_environment(
  name="local_linux",
  compatible_platforms=["linux_x86_64"],
  fallback_environment="local_busybox",
  ..
)

docker_environment(
  name="local_busybox",
  platform="linux_x86_64",
  image="busybox:latest@sha256-abcd123...",
  ..
)
```

### Environment-aware options

Environment targets have fields (target arguments) which correspond to options which are marked "environment-aware" (_TODO: align naming_). When an option is environment-aware, the value of the option that will be used in an environment can be overridden by setting the corresponding field value on the environment target for that environment. If an environment target does not set a value, it defaults to the value which is set globally via options values.

> For example, the [`[python-bootstrap].search_path` option](doc:reference-python-bootstrap#search_path) is environment-aware, which is indicated in its help (_TODO: ensure these are marked in the generated help_). It can be overridden for a particular environment by a corresponding environment target field, such as [the one on `local_environment`](doc:reference-local_environment#codepython_bootstrap_search_pathcode).

Reviewer: Environment-aware options indicate that they override options from the options system, but not vice-versa. Possibly worth adding a ticket for that.

---
title: "Environments: Cross-Platform or Remote Builds"
slug: "environments"
hidden: false
createdAt: "2022-10-03T21:39:51.235Z"
updatedAt: "2022-10-03T21:39:51.235Z"
---
Environments
============

By default, Pants will execute all sandboxed build work directly on localhost. But defining and using additional "environments" for particular targets allows Pants to transparently execute some or all of your build either:
1. locally in Docker containers
2. remotely via [remote execution](doc:remote-execution)

> 3. locally, but with a non-default set of environment variables and settings (such as for cross-building)

Reviewer 2: Another important use case is simply that macOS and Linux need different config. Environments can still be useful even without the 3 above use cases

## Defining environments

Environments are defined using environment targets:

* [`local_environment`](doc:reference-local_environment) - Runs without containerization on localhost (which is also the default if no environment targets are defined).

> * [`docker_environment`](doc:reference-docker_environment) - Runs in a cached container using the specified Docker image.

Reviewer 3: Does the `cached` here imply you have to `pull` any images outside of Pants before trying to use them?

Reviewer 4: Pants automatically attempts to pull images. "Cached" here means that Pants will spin up a single container for that image and then run multiple processes in that container for build actions (via essentially `docker exec`).

Reviewer: It might be worth specifying that "Docker" means "using the Docker installation on this machine"

* [`remote_environment`](doc:reference-remote_environment) - Runs in a remote worker via [remote execution](doc:remote-execution) (possibly with containerization, depending on the server implementation).

> Environment targets are given short, descriptive names using the [`[environments.names]` option](doc:reference-environments#names) (usually defined in `pants.toml`), which consuming targets use to refer to them in `BUILD` files. That might look like a `pants.toml` section and `BUILD` file (at the root of the repository in this case) containing:

Reviewer: ```suggestion
Give your environment targets short, descriptive names using the [`[environments.names]` option](doc:reference-environments#names) (usually defined in `pants.toml`), which consuming targets use to refer to them in `BUILD` files. That might look like a `pants.toml` section and `BUILD` file (at the root of the repository in this case) containing:
```

Make it clear that the "short descriptive names" are the responsibility of the reader

> ```toml

Reviewer 2: ```suggestion
```toml pants.toml
```

[environments.names]
linux = "//:local_linux"
linux_docker = "//:local_busybox"
```

> 

Reviewer 2: If you remove this blank line, then the code sample shows up as one, using tabs
```suggestion
```

> ```python

Reviewer 2: ```suggestion
```python BUILD
```

local_environment(
  name="local_linux",
  compatible_platforms=["linux_x86_64"],
  fallback_environment="local_busybox",
  ..
)

docker_environment(
  name="local_busybox",
  platform="linux_x86_64",
  image="busybox:latest@sha256-abcd123...",
  ..
)
```

### Environment-aware options

> Environment targets have fields (target arguments) which correspond to options which are marked "environment-aware". When an option is environment-aware, the value of the option that will be used in an environment can be overridden by setting the corresponding field value on the environment target for that environment. If an environment target does not set a value, it defaults to the value which is set globally via options values.

Author: See https://github.com/pantsbuild/pants/issues/17108 regarding making this section accurate.

Reviewer: Will be solved by way of #17124

> Environment targets have fields (target arguments) which correspond to options which are marked "environment-aware". When an option is environment-aware, the value of the option that will be used in an environment can be overridden by setting the corresponding field value on the environment target for that environment. If an environment target does not set a value, it defaults to the value which is set globally via options values.

Reviewer 3: I feel like the middle sentence here uses "environment" too many times, made my head spin. Maybe:
```suggestion
Environment targets have fields (target arguments) which correspond to options which are marked "environment-aware". When an option is environment-aware, the value of the option that will be used in an environment can be overridden by setting the corresponding field value on the linked environment target. If an environment target does not set a value, it defaults to the value which is set globally via options values.
```

> Environment targets have fields (target arguments) which correspond to options which are marked "environment-aware". When an option is environment-aware, the value of the option that will be used in an environment can be overridden by setting the corresponding field value on the environment target for that environment. If an environment target does not set a value, it defaults to the value which is set globally via options values.

Reviewer: "via options values" is a a hard thing to understand. Either link to a relevant page, or spell out "via pants.toml, environment variables when executing Pants, or command line options"/

Author: I'll make both `target` and `options` links here to the relevant pages.

> For example, the [`[python-bootstrap].search_path` option](doc:reference-python-bootstrap#search_path) is environment-aware, which is indicated in its help. It can be overridden for a particular environment by a corresponding environment target field, such as [the one on `local_environment`](doc:reference-local_environment#codepython_bootstrap_search_pathcode).

Reviewer 2: It would probably help to have a `BUILD` file example where you set this, and then have a comment in it saying "this overrides the option `[python-bootstrap].search_path`

People tend to really benefit from examples

> 👍 See an option which should be environment-aware, but isn't?
>
> Environments are a new concept: if you see an option value which should be marked environment-aware but isn't, please definitely [file an issue](https://github.com/pantsbuild/pants/issues/new/choose)!

## Consuming environments

> To declare which environment they should build with, many target types (but particularly "root" targets like tests or binaries) have an `environment=` field: for example, [`python_tests(environment=..)`](doc:reference-python_tests#codeenvironmentcode).

Reviewer 2: Is this sentence going to soon be stale?

Author: Unclear: we'll learn more during the alpha and rcs I expect.

The `environment=` field may either:
1. refer to an environment by name
2. use a special `__local__` environment name, which resolves to any matching `local_environment` (see "Environment matching" below)

Test targets additionally have a `runtime_environment=` field (_TODO: see workflow example below, and implement_) which defaults to the value of the target's `environment=` field, but which can be set explicitly to indicate that a test should execute in a different environment than it was built in. This can be used to enable cross-building (where a test is built on one platform, but executed on another), or to explicitly provide tools or running services at test runtime which would not otherwise be available.

> 🚧 Environment compatibility
> 

> > Currently, there is no static validation that a target's environment is compatible with its dependencies' environments -- only the implicit validation of the goals that you run successfully against those targets (`check`, `lint`, `test`, `package`, etc).

Reviewer: Do we provide static validation of runtime configurations when we're not using environments? If not, then this is probably likely to be more confusing than illuminating.

Author: We do for `resolve`s, `interpreter_constraints`, etc. So yea, sortof.

It a particular environment target _doesn't_ match, it can configure a `fallback_environment=` which will be attempted next. This allows for forming preference chains which are referred to by whichever environment name is at the head of the chain.

For example: a chain like "prefer remote execution if enabled, but fall back to local execution if the platform matches, otherwise use docker" might be configured via the targets:
```python
remote_environment(
  name="remote",
  fallback_environment="local",
  ..
)

local_environment(
  name="local",
  compatible_platforms=["linux_x86_64"],
  fallback_environment="docker",
)

docker_environment(
  name="docker",
  ..
)
```

In future versions, environment targets will gain additional predicates to control whether they match (for example: `local_environment` will likely gain a predicate that looks for the [presence or value of an environment variable](https://github.com/pantsbuild/pants/issues/17107). But in the meantime, it's possible to override which environments are matched for particular use cases by overriding their configured names: see the "Toggle use of an environment" workflow below for an example.

> ## Example workflows

Reviewer 2: Great idea

### Enabling remote execution globally

> `remote_environment` targets match unless the [`--remote-execution`](doc:reference-global#remote_execution) option is disabled. So to cause a particular environment name to use remote execution whenever it is enabled, you could define environment targets which tried remote execution first, and then fell back to local execution:

Author: This is about half accurate: making it accurate by introducing a deprecation of the current behavior to resolve https://github.com/pantsbuild/pants/issues/11148 will be my first goal tomorrow.

Author: Ok, nevermind: it is currently accurate, and so #11148 is resolved! I will look into whether we need a deprecation, or at the very least a backwards-compatibility note somewhere in this doc.

> 🚧 Speculation of remote execution
> 
> The `2.15.x` series of Pants does not yet support ["speculating" remote execution](https://github.com/pantsbuild/pants/issues/8353) by racing it against another environment (usually local or docker). While we expect that this will be necessary to make remote execution a viable option for local execution on user's laptops (where network connections are less reliable), it is less critical for CI use-cases.

### Use a `docker_environment` to build the inputs to a `docker_image`

To build a `docker_image` target containing a `pex_binary` which uses native (i.e. compiled) dependencies on a `macOS` machine, you can configure the `pex_binary` to be built in a `docker_environment`.

You'll need a `docker_environment` which uses an image containing the relevant build-time requirements of your PEX. At a minimum, you'll need Python itself:
```python
docker_environment(
  name="python_bullseye",
  platform="linux_x86_64",
  image="python:3.9.14-slim-bullseye@sha256-abcd123...",
  ..
)
```

> Next, mark your `pex_binary` target with this environment (with the name `python_bullseye`: see "Defining environments" above), and define a `docker_image` target depending on it.

Reviewer 2: `name` is ambiguous because it could refer to the "target name". I recommend being explicit saying `[environments-preview].names`

Author: Mm, the "Defining environments" section talks about how to give them short/descriptive names. Will leave it.

```python
pex_binary(
  name="main",

>   environment="python_bullseye",

Reviewer 3: Is it possible to `parameterize` an `environment`? For this example specifically, I'm typically calling `./pants package` on the `docker_image` target and the PEX is an implementation detail (therefore a Linux platform is what I want), but sometimes I want to run a PEX directly so I'll `./pants package` it expecting a Mac-compatible output.

Reviewer 3: I guess that is also achievable by overriding the environment name to point at my local env when I'm building the pex directly 🤔

Reviewer: It can be, and we should probably document that.

Author: Added a section.

)

docker_image(
    name="docker_image",
    instructions=[
        "FROM python:3.9.14-slim-bullseye@sha256-abcd123...",
        "ENTRYPOINT ["/main"]",
        "COPY examples/main.pex /main",
    ],
)
```

> 👍 Compatibility of `docker_environment` and `docker_image`
>
> Note that the Docker image used in your `docker_environment` does not need to match the base image of the `docker_image` targets that consume them: they only need to be compatible. This is because execution of build steps in a `docker_environment` occurs in an anonymous container, and only the required inputs are provided to the `docker_image` build.
>
> This means that your `docker_environment` can include things like compilers or other tools relevant to your build, without needing to manually use multi-stage Docker builds.

### Execute a test in Docker, while natively cross-building it

_TODO: Like https://github.com/pantsbuild/pants/issues/15764, but for tests._

> _TODO: Give an example of using `environment=` vs `runtime_environment=` to use docker only for test execution, but not for building of thirdparty dependencies by using PEX to cross-build. This will require exposing options out of PEX to override (?) the target platform, and figuring out how that doesn't break `check` is an open question._

Author: So: I'm pretty sure that we should avoid stabilizing environments until we have figured out exactly how to implement a cross-build like this. Many other very useful workflows are enabled, but this one in particular feels like it requires a bit more design work.

This is additionally covered by https://github.com/pantsbuild/pants/issues/8384.

### Use a `docker_environment` to build the inputs to a `docker_image`

To build a `docker_image` target containing a `pex_binary` which uses native (i.e. compiled) dependencies on a `macOS` machine, you can configure the `pex_binary` to be built in a `docker_environment`.

You'll need a `docker_environment` which uses an image containing the relevant build-time requirements of your PEX. At a minimum, you'll need Python itself:
```python BUILD
docker_environment(
  name="python_bullseye",
  platform="linux_x86_64",
  image="python:3.9.14-slim-bullseye@sha256-abcd123...",
  ..
)
```

Next, mark your `pex_binary` target with this environment (with the name `python_bullseye`: see "Defining environments" above), and define a `docker_image` target depending on it.

```python BUILD
pex_binary(
  name="main",

>   environment="python_bullseye",

Reviewer 2: This should show the `pants.toml` imo. Generally, I recommend maybe adding `-tgt` to the target name for all these env targets in this page. It's way too easy to think that `environment="target name"` rather than name as defined in pants.toml

For example, the [`[python-bootstrap].search_path` option](doc:reference-python-bootstrap#search_path) is environment-aware, which is indicated in its help. It can be overridden for a particular environment by a corresponding environment target field, such as [the one on `local_environment`](doc:reference-local_environment#codepython_bootstrap_search_pathcode).

> 👍 See an option which should be environment-aware, but isn't?
>
> Environments are a new concept: if you see an option value which should be marked environment-aware but isn't, please definitely [file an issue](https://github.com/pantsbuild/pants/issues/new/choose)!

## Consuming environments

To declare which environment they should build with, many target types (but particularly "root" targets like tests or binaries) have an `environment=` field: for example, [`python_tests(environment=..)`](doc:reference-python_tests#codeenvironmentcode).

> The `environment=` field may either:

Author: I guess that it is a convention that I'm trying to start.

I'm not really sure that end users know what a "field" is otherwise... most users probably think of the arguments to targets as ... arguments.

> The `environment=` field may either:

Reviewer 5: Perhaps we could have a glossary page listing terminology and syntax for how we typically refer to them, such as target field names always being on the form of `field_name=`, if we also do that in messages from `pants` that would make it more clear with fewer words in many cases I believe.

### Building one target in multiple environments

If a target will always need to be built in multiple environments (rather than conditionally based on which user is building it: see the "Toggle use of an environment for some consumers" section), then you can use the [`parametrize` builtin](doc:targets#parametrizing-targets) for the `environment=` field. If you had two environments named `linux` and `macos`, that would look like:

```python BUILD
pex_binary(
  name="bin",
  environment=parametrize("linux", "macos"),
)
```

### Environment matching

A single environment name may end up referring to different environment targets on different physical machines, or with different global settings applied: this is known as environment "matching".

* `local_environment` targets will match if their `compatible_platforms=` field matches localhost's platform.
* `docker_environment` targets will match [if Docker is enabled](doc:reference-global#docker_execution), and if their `platform=` field is compatible with localhost's platform.
* `remote_environment` targets will match [if Remote execution is enabled](doc:reference-global#remote_execution).

> It a particular environment target _doesn't_ match, it can configure a `fallback_environment=` which will be attempted next. This allows for forming preference chains which are referred to by whichever environment name is at the head of the chain.

Author: > Does the `fallback_environment` use the name as configured in `pants.toml` or a target address?

The former: the `field` docs and error messages should make that clear though.
